### ISSUE TITLE
Remove the need to hardcode absolute paths for some entries in `.cabal/config`: support xdg home variables

### ISSUE THREAD

**Issue by KholdStare** [haskell/cabal#1857]
Hi!

I keep all my configuration files on github to have an easily reproducible environment on all my machines. I keep my `.cabal/config` file in there as well, and don't like the fact that some absolute paths have to be hardcoded (e.g. `remote-repo-cache`, `world-file` and `build-summary`). This is inconvenient if `$HOME` happens to be different on various machines (e.g. work vs home).

Some possible solutions (which could be wrong):
- Expand `~` in paths.
- Expand environment variables from the shell (perhaps even very specific ones)
- Allow usage of `$prefix` in all paths

If one of the above suggestions is worthwhile, I'd be up for implementing it. Just let me know :smile:

Thank you!

**Comment by chris-martin** [comment]
Would someone with a commit bit mind weighing in on whether a patch would be accepted for this?

**Comment by hvr** [comment]
I generally like the idea, but under the condition we do *not* open up support for arbitrary environment variables.

I'd concretely suggest to add support for a couple additional `$`-variables

- `$home` via `System.Directory.getHomeDirectory`  (if we want, we could support `~/` as sugar
- `$xdg_data` / `$xdg_config` / `$xdg_cache` (via `System.Directory.getXdgDirectory`)

which I believe should address most use-cases without bringing into scope an unknown set of new variable names.

The downside of allowing env-variables to affect the paths in the cabal config is that file-change-monitoring gets slighlty more complicated, but by using an explicitly enumerated set of input variables (i.e. `$home` /`$xdg_{data,config,cache}`) this remains simple enough.

**Comment by sboosali** [comment]
yeah, I agree w.r.t. a fixed of environment variables. and the `~` sugar.

one issue is which environment variables are chosen for Windows (which is far as I know, isn't part of the XDG BaseDirs spex). For example, the standard implementations between Haskell and Racket differ (see below). I'd stick with whatever `directory` does (or if `Cabal` has its own XDG utilities, those), but I wanted to mention it.

Haskell (`directory` package):

* `$XDG_CONFIG_HOME` — `%APPDATA%`
* `$XDG_DATA_HOME`   — `%APPDATA%`
* `$XDG_CACHE_HOME`  — `%LOCALAPPDATA%`

Racket (`basedir` package):

* `$XDG_CONFIG_HOME` — `%LOCALAPPDATA%`
* `$XDG_DATA_HOME`   — `%LOCALAPPDATA%`
* `$XDG_CACHE_HOME`  — `%TEMP%`

<https://docs.racket-lang.org/basedir/index.html>
<http://hackage.haskell.org/package/directory-1.3.3.2/docs/System-Directory.html#t:XdgDirectory>

**Comment by 23Skidoo** [comment]
#645 seems to be another duplicate of this ticket.

**Comment by merijn** [comment]
> The downside of allowing env-variables to affect the paths in the cabal config is that file-change-monitoring gets slighlty more complicated, but by using an explicitly enumerated set of input variables (i.e. `$home` /`$xdg_{data,config,cache}`) this remains simple enough.

Do any of the filepath fields in ~/.cabal/config affect the hash of packages? Or just the file-change-monitoring?

**Comment by Mistuke** [comment]
https://github.com/haskell/cabal/issues/4597 is also a duplicate of this. There was a large discussion there about the migration issues. But I think with `new-build` we can avoid that as it's a new thing anyway.

But the short story is on Windows we shouldn't use `%AppData%` for anything other than cabal config file because this is a synced directory.

**Comment by 23Skidoo** [comment]
What's a synced directory on Windows? Synced across all users or backed up in the cloud?

**Comment by Mistuke** [comment]
Synced across active directory. It's a roaming profile folder. So it's synced back to the Ad server on login and logout. The data cabal puts there are often not useful on another machine but the sheer size of this folder makes network user's login and logout very slow as it has to sync and eats away at your profile quota. This affects most corporate environments and universities. 

In newer versions of Windows this has been made more explicit in the folder resolves to a "Roaming" path to point this out more.

**Comment by 23Skidoo** [comment]
Thanks, it's clear now.

**Comment by Mikolaj** [comment]
This is related to #680 that is now being worked on.

**Comment by athas** [comment]
Would people be opposed to just supporting `$home`?  I don't think such a simple feature should be conflated with the general problem of variables.

**Comment by gbaz** [comment]
i think this is still needed, and just allowing a simple set of environment variables to be expanded would be good and straightforward.

**Comment by phadej** [comment]
>  just allowing a simple set of environment variables

Not enough. If people want to use XDG paths, the related environment variables are often not set, i.e. we are using default locations,

**Comment by gbaz** [comment]
I think this would definitely be a good issue to direct some attention to. Especially with the user-config diff stuff actually working, we have pathways that help to share configs in general much better, but the lack of variable support stands in the way of how nice it could be.
